File: JAVA/Pigeonhole Sort/StackPointerPigeonholeSort.java
import java.io.*;

public class StackPointerPigeonholeSort
{
    static class Rating
    {
        int u,m;float r;long t;
        Rating(int u,int m,float r,long t){this.u=u;this.m=m;this.r=r;this.t=t;}
        public String toString(){return u+","+m+","+r+","+t;}
    }

    static class Node{Rating v;Node nx,pr;Node(Rating v){this.v=v;}}

    Node top=null; int size=0;

    void push(Rating r)
    {
        Node n=new Node(r);
        if(top!=null){n.pr=top;top.nx=n;} 
        top=n; 
        size++;
    }

    void pigeonholeSort()
    {
        final int HOLES=10;

        Node base=top; while(base!=null&&base.pr!=null) base=base.pr;

        @SuppressWarnings("unchecked") java.util.ArrayDeque<Node>[] holes=new java.util.ArrayDeque[HOLES];
        for(int i=0;i<HOLES;i++) holes[i]=new java.util.ArrayDeque<>();

        for(Node p=base;p!=null;p=p.nx) holes[(int)Math.round(p.v.r*2)-1].addLast(p);

        top=null; size=0; Node prev=null;
        for(int h=0;h<HOLES;h++) while(!holes[h].isEmpty())
        {
            Node n=holes[h].pollFirst(); n.nx=null; n.pr=prev;
            if(prev!=null) prev.nx=n;
            prev=n; if(top==null) top=n; size++;
        }

        while(top.nx!=null) top=top.nx;
    }

    public static void main(String[]a)
    throws Exception
    {
        if(a.length<2)
        {
            System.out.println("Uso: java StackPointerPigeonholeSort <ratings.csv> <N>");
            return;
        }

        int N=Integer.parseInt(a[1]); StackPointerPigeonholeSort S=new StackPointerPigeonholeSort();

        try(BufferedReader br=new BufferedReader(new FileReader(a[0])))
        {
            String ln=br.readLine();int c=0;
            while((ln=br.readLine())!=null&&c<N)
            {
                String[] f=ln.split(",");
                S.push(new Rating(Integer.parseInt(f[0]),Integer.parseInt(f[1]),
                                  Float.parseFloat(f[2]),Long.parseLong(f[3])));
                c++;
            }
        }

        long t0=System.nanoTime();S.pigeonholeSort();long t1=System.nanoTime();
        System.out.printf("pilha ponteiro %d regs %.2f ms%n",N,(t1-t0)/1e6);

        Node b=S.top; while(b.pr!=null) b=b.pr;
        for(int i=0;i<20&&b!=null;i++,b=b.nx)System.out.println(b.v);
    }
}

File: Ruby/PigeonholeSort/stack_pointer_pigeonhole_sort.rb
require 'benchmark'
require 'objspace'

class Rating
  attr_accessor :user_id, :movie_id, :rating, :timestamp

  def initialize(user_id, movie_id, rating, timestamp)
    @user_id = user_id
    @movie_id = movie_id
    @rating = rating
    @timestamp = timestamp
  end

  def to_s
    "#{@user_id},#{@movie_id},#{@rating},#{@timestamp}"
  end
end

class Node
  attr_accessor :rating, :next, :prev

  def initialize(rating)
    @rating = rating
    @next = nil
    @prev = nil
  end
end

class StackPointerPigeonholeSort
  attr_accessor :top, :size

  def initialize
    @top = nil
    @size = 0
  end

  def push(rating)
    node = Node.new(rating)
    if @top.nil?
      @top = node
    else
      node.prev = @top
      @top.next = node
      @top = node
    end
    @size += 1
  end

  def pigeonhole_sort
    # Volta para o início da lista
    base = @top
    base = base.prev while base&.prev

    holes = Array.new(10) { [] }
    p = base
    while p
      idx = [(p.rating.rating * 2).round - 1, 9].min
      idx = [idx, 0].max
      holes[idx] << p
      p = p.next
    end

    @top = nil
    @size = 0
    prev = nil

    holes.each do |hole|
      hole.each do |node|
        node.next = nil
        node.prev = prev
        prev&.next = node
        prev = node
        @top = node if @top.nil?
        @size += 1
      end
    end

    @top = prev
  end
end

# ------------------- EXECUÇÃO PRINCIPAL -------------------

if ARGV.size < 2
  puts "Uso: ruby #{__FILE__} <arquivo.csv> <N>"
  exit
end

path = ARGV[0]
n = ARGV[1].to_i
pilha = StackPointerPigeonholeSort.new

File.open(path) do |f|
  f.readline  # pula cabeçalho
  f.each_line.take(n).each do |line|
    user_id, movie_id, rating, timestamp = line.strip.split(',')
    pilha.push(Rating.new(user_id.to_i, movie_id.to_i, rating.to_f, timestamp.to_i))
  end
end

mem_before = ObjectSpace.memsize_of_all
time = Benchmark.realtime { pilha.pigeonhole_sort }
mem_after = ObjectSpace.memsize_of_all

puts "Ordenado #{n} registros em #{(time * 1000).round(2)} ms"
puts "Memória usada: #{((mem_after - mem_before) / 1024.0).round(2)} KB"

puts "\nPrimeiros 100 registros ordenados:"
cur = pilha.top
cur = cur.prev while cur&.prev
100.times do
  break if cur.nil?
  puts cur.rating
  cur = cur.next
end

puts "\nÚltimos 100 registros ordenados:"
cur = pilha.top
100.times do
  break if cur.nil?
  puts cur.rating
  cur = cur.prev
end

puts "\nTamanho da pilha: #{pilha.size}"